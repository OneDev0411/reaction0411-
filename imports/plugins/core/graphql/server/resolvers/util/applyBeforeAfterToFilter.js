import { decodeOpaqueId } from "@reactioncommerce/reaction-graphql-xforms/id";

/**
 * @name applyBeforeAfterToFilter
 * @method
 * @memberof GraphQL/ResolverUtilities
 * @summary Adjusts a MongoDB filter based on GraphQL `before` and `after` params
 * @return {Promise<Object>} The potentially-modified filter object
 */
export default async function applyBeforeAfterToFilter({
  after,
  baseFilter = {},
  before,
  collection,
  sortBy: sortByField = "_id",
  sortOrder = "asc"
}) {
  let filter = baseFilter;
  const limits = {};
  const ors = [];

  if (typeof sortByField !== "string") throw new Error("sortBy is required");
  if (sortOrder !== "asc" && sortOrder !== "desc") throw new Error("sortOrder is required");

  if (after && before) throw new Error("Including both 'after' and 'before' params is not allowed");

  if (!after && !before) return filter;

  let encodedId;
  let op;
  if (before) {
    encodedId = before;
    op = (sortOrder === "desc" ? "$gt" : "$lt");
  } else {
    encodedId = after;
    op = (sortOrder === "desc" ? "$lt" : "$gt");
  }

  // "encodedId" is encoded twice. At this point the cursor has already been decoded
  // but we are still left with an encoded ID, which we'll now decode.
  let internalId;
  if (encodedId.endsWith("=")) {
    internalId = decodeOpaqueId(encodedId).id;
  } else {
    internalId = encodedId; // it was only encoded once
  }

  if (sortByField === "_id") {
    // We already have _id. Skip the lookup
    filter = {
      $and: [
        { ...filter },
        { _id: { [op]: internalId } }
      ]
    };
  } else {
    const doc = await collection.findOne({
      _id: internalId
    }, {
      fields: {
        [sortByField]: 1
      }
    });

    if (doc) {
      limits[op] = doc[sortByField];
      ors.push({
        [sortByField]: doc[sortByField],
        _id: { [op]: internalId }
      });

      filter = {
        $and: [
          { ...filter },
          {
            $or: [
              {
                [sortByField]: limits
              },
              ...ors
            ]
          }
        ]
      };
    }
  }

  return filter;
}
