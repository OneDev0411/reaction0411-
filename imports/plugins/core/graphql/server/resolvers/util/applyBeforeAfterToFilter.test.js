import applyBeforeAfterToFilter from "./applyBeforeAfterToFilter";

test("throws an error if sortBy is missing", () =>
  expect(applyBeforeAfterToFilter({ sortBy: null })).rejects.toMatchSnapshot());

test("throws an error if sortOrder is missing", () =>
  expect(applyBeforeAfterToFilter({ sortOrder: null })).rejects.toMatchSnapshot());

test("throws an error if both before and after are set", () =>
  expect(applyBeforeAfterToFilter({ after: "123", before: "123" })).rejects.toMatchSnapshot());

describe("after", () => {
  test("alters filter correctly when sorting by _id ascending (default)", async () => {
    const baseFilter = { _id: { $in: ["abc"] } };
    const result = await applyBeforeAfterToFilter({ after: "cmVhY3Rpb24vc2hvcDoxMjM=", baseFilter });

    expect(result).toEqual({
      $and: [
        baseFilter,
        { _id: { $gt: "123" } }
      ]
    });
  });

  test("alters filter correctly when sorting by _id descending", async () => {
    const baseFilter = { _id: { $in: ["abc"] } };
    const result = await applyBeforeAfterToFilter({ after: "cmVhY3Rpb24vc2hvcDoxMjM=", baseFilter, sortOrder: "desc" });

    expect(result).toEqual({
      $and: [
        baseFilter,
        { _id: { $lt: "123" } }
      ]
    });
  });

  test("alters filter correctly when sorting by a non-_id field ascending", async () => {
    const baseFilter = { _id: { $in: ["abc"] } };
    const name = "Some Name";
    const result = await applyBeforeAfterToFilter({
      after: "cmVhY3Rpb24vc2hvcDoxMjM=",
      baseFilter,
      collection: {
        findOne: () => ({ name })
      },
      sortBy: "name"
    });

    expect(result).toEqual({
      $and: [
        baseFilter,
        {
          $or: [
            { name: { $gt: name } },
            { _id: { $gt: "123" }, name }
          ]
        }
      ]
    });
  });

  test("alters filter correctly when sorting by a non-_id field descending", async () => {
    const baseFilter = { _id: { $in: ["abc"] } };
    const name = "Some Name";
    const result = await applyBeforeAfterToFilter({
      after: "cmVhY3Rpb24vc2hvcDoxMjM=",
      baseFilter,
      collection: {
        findOne: () => ({ name })
      },
      sortBy: "name",
      sortOrder: "desc"
    });

    expect(result).toEqual({
      $and: [
        baseFilter,
        {
          $or: [
            { name: { $lt: name } },
            { _id: { $lt: "123" }, name }
          ]
        }
      ]
    });
  });
});

describe("before", () => {
  test("alters filter correctly when sorting by _id ascending (default)", async () => {
    const baseFilter = { _id: { $in: ["abc"] } };
    const result = await applyBeforeAfterToFilter({ before: "cmVhY3Rpb24vc2hvcDoxMjM=", baseFilter });

    expect(result).toEqual({
      $and: [
        baseFilter,
        { _id: { $lt: "123" } }
      ]
    });
  });

  test("alters filter correctly when sorting by _id descending", async () => {
    const baseFilter = { _id: { $in: ["abc"] } };
    const result = await applyBeforeAfterToFilter({ before: "cmVhY3Rpb24vc2hvcDoxMjM=", baseFilter, sortOrder: "desc" });

    expect(result).toEqual({
      $and: [
        baseFilter,
        { _id: { $gt: "123" } }
      ]
    });
  });

  test("alters filter correctly when sorting by a non-_id field ascending", async () => {
    const baseFilter = { _id: { $in: ["abc"] } };
    const name = "Some Name";
    const result = await applyBeforeAfterToFilter({
      before: "cmVhY3Rpb24vc2hvcDoxMjM=",
      baseFilter,
      collection: {
        findOne: () => ({ name })
      },
      sortBy: "name"
    });

    expect(result).toEqual({
      $and: [
        baseFilter,
        {
          $or: [
            { name: { $lt: name } },
            { _id: { $lt: "123" }, name }
          ]
        }
      ]
    });
  });

  test("alters filter correctly when sorting by a non-_id field descending", async () => {
    const baseFilter = { _id: { $in: ["abc"] } };
    const name = "Some Name";
    const result = await applyBeforeAfterToFilter({
      before: "cmVhY3Rpb24vc2hvcDoxMjM=",
      baseFilter,
      collection: {
        findOne: () => ({ name })
      },
      sortBy: "name",
      sortOrder: "desc"
    });

    expect(result).toEqual({
      $and: [
        baseFilter,
        {
          $or: [
            { name: { $gt: name } },
            { _id: { $gt: "123" }, name }
          ]
        }
      ]
    });
  });
});
