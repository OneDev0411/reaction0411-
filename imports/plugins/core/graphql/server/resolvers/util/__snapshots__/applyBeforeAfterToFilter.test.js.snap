// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`throws an error if both before and after are set 1`] = `[Error: Including both 'after' and 'before' params is not allowed]`;

exports[`throws an error if sortBy is missing 1`] = `[Error: sortBy is required]`;

exports[`throws an error if sortOrder is missing 1`] = `[Error: sortOrder is required]`;
